protocol/contracts/reserve/src/contract.rs

```diff
@@ -0,0 +1,118 @@
+use std::ops::DerefMut;
+
+use access_control::SingleUserAccess;
+use finance::coin::Coin;
+use platform::{
+    bank::{self, BankAccount, BankAccountView},
+    batch::Batch,
+    contract, error as platform_error,
+    message::Response as PlatformResponse,
+    response,
+};
+use sdk::{
+    cosmwasm_ext::Response as CwResponse,
+    cosmwasm_std::{self, entry_point, Binary, Deps, DepsMut, Env, MessageInfo},
+};
+use versioning::{package_version, version, SemVer, Version, VersionSegment};
+
+use crate::{
+    api::{ConfigResponse, ExecuteMsg, InstantiateMsg, LpnCurrency, MigrateMsg, QueryMsg},
+    error::{ContractError, ContractResult},
+    state::Config,
+};
+
+// const CONTRACT_STORAGE_VERSION_FROM: VersionSegment = 0;
+const CONTRACT_STORAGE_VERSION: VersionSegment = 0;
+const PACKAGE_VERSION: SemVer = package_version!();
+const CONTRACT_VERSION: Version = version!(CONTRACT_STORAGE_VERSION, PACKAGE_VERSION);
+
+#[entry_point]
+pub fn instantiate(
+    mut deps: DepsMut<'_>,
+    _env: Env,
+    _info: MessageInfo,
+    new_reserve: InstantiateMsg,
+) -> ContractResult<CwResponse> {
+    deps.api
+        .addr_validate(new_reserve.lease_code_admin.as_str())
+        .map_err(Into::<ContractError>::into)
+        .and_then(|_| {
+            SingleUserAccess::new(
+                deps.storage.deref_mut(),
+                crate::access_control::LEASE_CODE_ADMIN_KEY,
+            )
+            .grant_to(&new_reserve.lease_code_admin)
+            .map_err(Into::into)
+        })
+        .and_then(|()| versioning::initialize(deps.storage, CONTRACT_VERSION).map_err(Into::into))
+        .and_then(|()| Config::from(new_reserve).store(deps.storage))
+        .map(|()| response::empty_response())
+        .or_else(|err| platform_error::log(err, deps.api))
+}
+
+#[entry_point]
+pub fn migrate(deps: DepsMut<'_>, __env: Env, _msg: MigrateMsg) -> ContractResult<CwResponse> {
+    versioning::update_software(deps.storage, CONTRACT_VERSION, Into::into)
+        .and_then(response::response)
+        .or_else(|err| platform_error::log(err, deps.api))
+}
+
+#[entry_point]
+pub fn execute(
+    mut deps: DepsMut<'_>,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> ContractResult<CwResponse> {
+    match msg {
+        ExecuteMsg::NewLeaseCode {
+            code_id: new_code_id,
+        } => {
+            SingleUserAccess::new(
+                deps.storage.deref_mut(),
+                crate::access_control::LEASE_CODE_ADMIN_KEY,
+            )
+            .check(&info.sender)?;
+
+            Config::update_lease_code(deps.storage, new_code_id.into())
+                .map(|()| PlatformResponse::default())
+        }
+        ExecuteMsg::CoverLiquidationLosses { amount } => {
+            let lease = info.sender;
```

Comment:
not sure I understand your point

Reply:
I meant that we're assigning `lease` before we check that it actually is a lease /+85/. While it is true that right now all the errors are chained logically with AND, it might not always be the case, e.g. introduction of `Result::or_else` calls.

Reply:
here it's named for convenience, just a shorter name, ... that does not change the semantic

Reply:
I don't have beef with the name, I wanted to emphasize that usually when we assign it it's usually the version of the value we'd like to use and is usually already checked.

Feel free to close this comment if you don't agree. It's just a minor concern of mine.

Reply:
no, I see merit in it .... though the safer alternative, passing it as function arg, would be more complex